bauble_macro_util/src/lib.rs

```diff
@@ -0,0 +1,833 @@
+use std::collections::HashSet;
+
+use proc_macro2::{Ident, Span, TokenStream};
+use quote::quote;
+use syn::{
+    parse2, spanned::Spanned, AttrStyle, Attribute, Data, DeriveInput, Error, Expr, Fields,
+    ImplGenerics, Index, Token, Type, WhereClause,
+};
+
+// Related fields used by `derive_struct` and `derive_fields` containing type info
+struct TypeInfo<'a> {
+    // The struct or variant, used for construction
+    ty: proc_macro2::TokenStream,
+    // The type's generics
+    impl_generics: &'a ImplGenerics<'a>,
+    where_clause: Option<&'a WhereClause>,
+}
+
+// Generate code to deserialize a struct or variant with fields
+fn derive_fields(
+    TypeInfo {
+        ty,
+        impl_generics,
+        where_clause,
+    }: TypeInfo,
+    // The struct or variant's fields
+    fields: &Fields,
+    // Whether the struct or variant should be parsed from a tuple. For structs with named
+    // fields, this is the case if it has the `tuple` attribute
+    tuple: bool,
+) -> proc_macro2::TokenStream {
+    // General kind of field
+    enum FieldTy<'a> {
```

Comment:
Maybe move this to the outer scope and change all the comments on it to doc comments? Seems like a waste to not use such nice documentation if it already exists.